protocol/communities/community_event.go

```diff
@@ -0,0 +1,430 @@
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_CATEGORY_EDIT:
+		_, err := o.editCategory(communityEvent.CategoryData.CategoryId, communityEvent.CategoryData.Name, communityEvent.CategoryData.ChannelsIds)
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_CHANNEL_CREATE:
+		err := o.createChat(communityEvent.ChannelData.ChannelId, communityEvent.ChannelData.Channel)
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_CHANNEL_DELETE:
+		o.deleteChat(communityEvent.ChannelData.ChannelId)
+
+	case protobuf.CommunityEvent_COMMUNITY_CHANNEL_EDIT:
+		err := o.editChat(communityEvent.ChannelData.ChannelId, communityEvent.ChannelData.Channel)
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_CHANNEL_REORDER:
+		_, err := o.reorderChat(communityEvent.ChannelData.CategoryId, communityEvent.ChannelData.ChannelId, int(communityEvent.ChannelData.Position))
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_CATEGORY_REORDER:
+		_, err := o.reorderCategories(communityEvent.CategoryData.CategoryId, int(communityEvent.CategoryData.Position))
+		if err != nil {
+			return err
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT:
+		for pkString, addedMember := range communityEvent.MembersAdded {
+			pk, err := common.HexToPubkey(pkString)
+			if err != nil {
+				return err
+			}
+			if !o.HasMember(pk) {
+				o.addCommunityMember(pk, addedMember)
+			}
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT:
+		for pkString := range communityEvent.RejectedRequestsToJoin {
+			pk, err := common.HexToPubkey(pkString)
+			if err != nil {
+				return err
+			}
+			o.removeMemberFromOrg(pk)
+		}
+
+	case protobuf.CommunityEvent_COMMUNITY_MEMBER_KICK:
+		pk, err := common.HexToPubkey(communityEvent.MemberToAction)
+		if err != nil {
+			return err
+		}
+
+		if o.IsMemberOwnerOrAdmin(pk) {
+			return errors.New("attempt to kick an owner or admin of the community from the admin side")
+		}
+
+		o.removeMemberFromOrg(pk)
+
+	case protobuf.CommunityEvent_COMMUNITY_MEMBER_BAN:
+		pk, err := common.HexToPubkey(communityEvent.MemberToAction)
+		if err != nil {
+			return err
+		}
+
+		if o.IsMemberOwnerOrAdmin(pk) {
+			return errors.New("attempt to ban an owner or admin of the community from the admin side")
+		}
+		o.banUserFromCommunity(pk)
+
+	case protobuf.CommunityEvent_COMMUNITY_MEMBER_UNBAN:
+		pk, err := common.HexToPubkey(communityEvent.MemberToAction)
+		if err != nil {
+			return err
+		}
+		o.unbanUserFromCommunity(pk)
+
+	default:
+		return errors.New("unknown admin community event")
+	}
+	return nil
+}
+
+func (o *Community) NewCommunityEventClock() uint64 {
+	if len(o.config.Events) == 0 {
+		return o.config.CommunityDescription.Clock + 1
+	}
+	return o.config.Events[len(o.config.Events)-1].CommunityEventClock + 1
```

Comment:
I think you want to use a `timestamp` as a seed, just like clock, rather than a straight lamport timestamp, similarly as we do with Clock values (it helps when you make changes and you might not have the latest state)

Reply:
I see, it's not super important, we can keep a straight lamport timestamp, but we'll have to break ties.

Though not sure about the last sentence "Also, it allows me to track that we got all admin updates and apply admins events to the CommunityDescription"
Basically we can't ensure that we have all admin events, and we should not expect events to be strictly in sequence, in group chat we piggy back the history, so that's probably something we should be consider doing here as well.

For example: 
1) admin-1 sends event admin-1-event
2) admin-2 receives event admin-1-event
3) admin-2 sends event admin-2-event

In this case, admin-2 will also publish together admin-1-event & admin-2-event, therefore preserving the history.
Once the owner collates the events, that won't be necessary anymore.
That's an option worth considering.

Reply:
Yes, the current logic is that admin-2  will send admin-1-event & send admin-2-event
But I just realized another problem:
1. admin-1 sends event admin-1-event
2. admin-2 receives event admin-1-event
3. admin-2 sends event admin-2-event

Owner accepts admin-1 event. He certifies it and publishes updated CommunityDescription with updated Clock
Admin-2 event will be rejected as the owner increased the CommunityDescription clock.

It seems we should not bind admin events to the CommunityDescription clock at all

I need to think about the solution

Reply:
I think admin events should probably be tied to a specific version of a community description (which could be identiified by a clock value), but as you mentioned it should not be bound to it